### PR TITLE
DNM: rgw zipper wip

### PR DESCRIPTION
**Pull request by kalebskeithley**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/CMakeLists.txt [hunk 1]
```diff
@@ -372,6 +310,214 @@ set_target_properties(radosgw PROPERTIES OUTPUT_NAME radosgw VERSION 2.0.0
...
+  services/svc_user_rados.cc
+  services/svc_zone.cc
+  services/svc_zone_utils.cc
+  rgw_bucket.cc
+  rgw_cr_rados.cc
+  rgw_d3n_datacache.cc
+  rgw_data_sync.cc
+  rgw_gc.cc
+  rgw_gc_log.cc
+  rgw_lua_request.cc
```
Comment (mattbenjamin) [inline src/rgw/CMakeLists.txt:346]: rgw_lua_request.cc shouldn't be rados specific?

File: src/rgw/CMakeLists.txt [hunk 2]
```diff
@@ -372,6 +310,214 @@ set_target_properties(radosgw PROPERTIES OUTPUT_NAME radosgw VERSION 2.0.0
...
+  rgw_reshard.cc
+  rgw_rest_bucket.cc
+  rgw_rest_log.cc
+  rgw_rest_pubsub.cc
+  rgw_rest_pubsub_common.cc
+  rgw_rest_realm.cc
+  rgw_rest_s3.cc
+  rgw_sal_rados.cc
+  rgw_service.cc
+  rgw_sync.cc
```
Comment (mattbenjamin) [inline src/rgw/CMakeLists.txt:364]: until future refactoring, I guess keeping all of sync below the line makes sense

File: src/rgw/rgw_sal.cc [hunk 3]
```diff
@@ -83,161 +72,61 @@ RGWObjState::RGWObjState(const RGWObjState& rhs) : obj (rhs.obj) {
...
-      delete store;
-      return nullptr;
-    }
-    return store;
+  const char *dlname = "/usr/lib64/ceph/librgw_sal_rados.so";
+  rgw::sal::Store* store = nullptr;
+  void *dl = nullptr;
+  rgw::sal::Store *(*newStore)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool) = nullptr;
+  if (svc.compare("d3n") == 0) {
+    dlname = "/usr/lib64/ceph/librgw_sal_d3n.so";
```
Comment (mattbenjamin) [inline src/rgw/rgw_sal.cc:80]: can we make this work without full paths here?

File: src/rgw/CMakeLists.txt [hunk 4]
```diff
@@ -57,112 +30,78 @@ set(librgw_common_srcs
...
   rgw_compression.cc
+  rgw_coroutine.cc
+  rgw_cr_rest.cc
+  rgw_cr_tools.cc
+  rgw_crypt.cc
+  rgw_crypt_sanitize.cc
   rgw_etag_verifier.cc
   rgw_cors.cc
   rgw_cors_s3.cc
+  rgw_datalog.cc
```
Comment (dang) [inline src/rgw/CMakeLists.txt:47]: datalog has cls stuff in it, so it should be in radosstore

File: src/rgw/CMakeLists.txt [hunk 5]
```diff
@@ -57,112 +30,78 @@ set(librgw_common_srcs
...
+  rgw_json_enc.cc
   rgw_keystone.cc
+  rgw_kmip_client.cc
+  rgw_kms.cc
   rgw_ldap.cc
   rgw_lc.cc
   rgw_lc_s3.cc
-  rgw_lc_tier.cc
+  rgw_lc_tier.cc 
+  rgw_log_backing.cc
```
Comment (dang) [inline src/rgw/CMakeLists.txt:62]: This is CLS stuff too.

File: src/rgw/CMakeLists.txt [hunk 6]
```diff
@@ -57,112 +30,78 @@ set(librgw_common_srcs
...
+  rgw_lua.cc
+  rgw_lua_utils.cc
   rgw_metadata.cc
```
Comment (dang) [inline src/rgw/CMakeLists.txt:65]: metadata log is CLS, currently

File: src/rgw/CMakeLists.txt [hunk 7]
```diff
@@ -57,112 +30,78 @@ set(librgw_common_srcs
...
-  rgw_s3select.cc
+  rgw_rest_sts.cc
+  rgw_rest_user_policy.cc
   rgw_role.cc
+  rgw_s3select.cc
   rgw_sal.cc
-  rgw_sal_rados.cc
   rgw_string.cc
+  rgw_sts.cc
+  rgw_sync_error_repo.cc
```
Comment (dang) [inline src/rgw/CMakeLists.txt:91]: sync stuff is below the light right now.

File: src/rgw/CMakeLists.txt [hunk 8]
```diff
@@ -57,112 +30,78 @@ set(librgw_common_srcs
...
+  rgw_sync_policy.cc
```
Comment (dang) [inline src/rgw/CMakeLists.txt:92]: sync stuff is below the line

File: src/rgw/CMakeLists.txt [hunk 9]
```diff
@@ -57,112 +30,78 @@ set(librgw_common_srcs
...
-  rgw_kmip_client.cc
-  rgw_url.cc
-  rgw_oidc_provider.cc
-  rgw_datalog.cc
-  cls_fifo_legacy.cc
-  rgw_lua_utils.cc
-  rgw_lua.cc
-  rgw_bucket_encryption.cc
-  rgw_tracer.cc)
+  cls_fifo_legacy.cc)
```
Comment (dang) [inline src/rgw/CMakeLists.txt:104]: cls is below the line.

File: src/rgw/CMakeLists.txt [hunk 10]
```diff
@@ -269,32 +208,31 @@ endif()
...
   rgw_http_client_curl.cc
   rgw_loadgen.cc
   rgw_log.cc
   rgw_lua_request.cc
+  rgw_opa.cc
+  rgw_os_lib.cc
   rgw_period_pusher.cc
+  rgw_process.cc
   rgw_realm_reloader.cc
   rgw_realm_watcher.cc
```
Comment (dang) [inline src/rgw/CMakeLists.txt:223]: Hmm... I had realm_watcher below, but maybe it goes with real_reloader?  It's only being created for RadosStore currently, but it's not conditionally compiled.

**Review by dang — COMMENTED** [review]
Just ran through CMakeLists.txt.  The comments bring it up to where I left of in my work on modules.  It's not complete, though.

File: src/rgw/rgw_sal.cc (hunk 3 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal.cc:80]: Eventually I expect this to come from whatever the config solution is.  So yes.

File: src/rgw/CMakeLists.txt (hunk 8 above)
Comment (kalebskeithley) [reply src/rgw/CMakeLists.txt:92]: there are calls to rgw_sync_pipe_filter::{encode,decode} in rgw_common.cc, rgw_zone.cc, and dbstore/sqlite/sqliteDB.cc
there are calls to rgw_sync_policy_info::{decode_json,dump} in rgw_common.cc and rgw_zone.cc.
there is a calle to rgw_sync_policy_info::dump in rgw_admin.cc

Are these a layering violation, or ???

Comment (kalebskeithley) [reply src/rgw/CMakeLists.txt:92]: There are calls in rgw_common.cc and rgw_zone.cc to rgw_sync_pipe_filter::{encode,decode}()
There is a call in store/dbstore/sqliteDB.cc to rgw_sync_pipe_filter::decode()
There are calls in rgw_common.cc, rgw_zone.cc, and rgw_sync_policy_info::{decode_json,dump}()
There is a call in grw_admin.cc to rgw_sync_policy_info::dump()

These turn into undefined refs after moving rgw_sync_policy.cc below the line.

**Pull request closed without merging**